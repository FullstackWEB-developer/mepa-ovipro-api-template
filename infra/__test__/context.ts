import { Project } from '@alma-cdk/project';
import { Construct } from 'constructs';
import projectProps from '../bin/config';

class TestContextProvider extends Construct {
    constructor(scope: Construct, id: string) {
        super(scope, id);
        this.node.setContext('account', 'mock');
        this.node.setContext('environment', 'mock1');
    }
}

/**Creates a CDK construct with all the required context fro running tests and assignes it into a CDK app. */
export function createCdkTestContext(): Construct {
    const app = new Project(projectProps);
    const testContextProvider = new TestContextProvider(app, 'TestContextProvider');
    return testContextProvider;
}
