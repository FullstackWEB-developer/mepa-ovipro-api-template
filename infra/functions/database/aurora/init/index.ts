import AWS from 'aws-sdk';
import {
    CloudFormationCustomResourceEvent,
    CloudFormationCustomResourceDeleteEvent,
    CloudFormationCustomResourceCreateEvent,
    CloudFormationCustomResourceUpdateEvent,
    CloudFormationCustomResourceResponse,
} from 'aws-lambda';

const rdsDataService = new AWS.RDSDataService();
const secretsmanager = new AWS.SecretsManager();

const createCreateDatabaseQueries = () => [
    `create database ovipro;`,
    `comment on database ovipro is 'Ovipro default database';`,
];

/**
 * Creates users into the database. User details are fetched from Secret manager.
 * Clone (suffix _clone) users are created for the Flyway user with createrole permission to support
 * secret rotation properly.
 */
const createFlywayUserQueries = async () => {
    const flywayParams = {
        SecretId: process.env.FLYWAY_USER_SECRET_NAME || '',
        VersionStage: 'AWSCURRENT',
    };

    const oviproParams = {
        SecretId: process.env.READ_WRITE_USER_SECRET_NAME || '',
        VersionStage: 'AWSCURRENT',
    };

    const flywaySecret = await secretsmanager.getSecretValue(flywayParams).promise();

    const flywaySecretString = JSON.parse(flywaySecret.SecretString || '');

    const oviproSecret = await secretsmanager.getSecretValue(oviproParams).promise();

    const oviproSecretString = JSON.parse(oviproSecret.SecretString || '');

    // Multi-user rotation Lambda creates clone users during password/user rotation, but it doesn't include
    // createrole permission, which we want Flyway to have. Thus we create flyway_user and flyway_user_clone
    // manually with correct grants.
    return [
        `create user ${flywaySecretString.username} with encrypted password '${flywaySecretString.password}' createrole;`,
        `create user ${flywaySecretString.username}_clone with encrypted password '${flywaySecretString.password}' createrole;`,
        `create user ${oviproSecretString.username} with encrypted password '${oviproSecretString.password}';`,
        `grant all privileges on database ovipro to ${flywaySecretString.username};`,
        `grant all privileges on database ovipro to ${flywaySecretString.username}_clone;`,
        `grant ${oviproSecretString.username} to ${flywaySecretString.username};`,
        `grant ${oviproSecretString.username} to ${flywaySecretString.username}_clone;`,
    ];
};

/**
 * Executes given query to database
 *
 * @param query
 * @param transactionId
 */
const executeStatement = async (query: string) => {
    const sqlParams: AWS.RDSDataService.ExecuteStatementRequest = {
        secretArn: process.env.DB_SECRET_ARN || '',
        resourceArn: process.env.DB_CLUSTER_ARN || '',
        sql: query,
        database: process.env.DB_NAME,
        includeResultMetadata: true,
    };

    try {
        const result = await rdsDataService.executeStatement(sqlParams).promise();
        console.info(`executed query, result: ${JSON.stringify(result)}`);
    } catch (e) {
        console.error(`could not execute statement, error: ${e}`);
    }
};

/**
 * Reads file from given path and executes queries inside
 *
 * @param filePath
 */
const executeQueries = async () => {
    const flywayUserQueries = await createFlywayUserQueries();
    const queries = [...createCreateDatabaseQueries(), ...flywayUserQueries];
    try {
        /**
         * Cant use Promise.all here, we want to execute in series
         */
        for (const query of queries) {
            await executeStatement(query);
        }
    } catch (e) {
        console.error(`could execute statement, error: ${e}`);
    }
};

/**
 * Run migrations scripts when event is CREATE
 *
 * NOTE: currently handles also UPDATE-events
 *
 * @param event
 * @returns cfnresponse
 */
const onCreate = async (
    event: CloudFormationCustomResourceCreateEvent | CloudFormationCustomResourceUpdateEvent,
): Promise<CloudFormationCustomResourceResponse> => {
    console.info(`create new resource with props ${JSON.stringify(event.ResourceProperties)}`);
    console.info(`database arn: ${process.env.DB_SECRET_ARN || ''}`);
    console.info(`cluster arn: ${process.env.DB_CLUSTER_ARN || ''}`);
    console.info(`database name: ${process.env.DB_NAME || ''}`);
    // If multiple resources are created this cant be hardcoded
    const physicalId = 'DatabaseInitMigrationResource';

    try {
        console.info({ event });

        await executeQueries();

        console.info('migration complete');

        return {
            PhysicalResourceId: physicalId,
            Status: 'SUCCESS',
            RequestId: event.RequestId,
            StackId: event.StackId,
            LogicalResourceId: event.LogicalResourceId,
            Reason: `created database and user for migrations`,
        };
    } catch (e) {
        return {
            PhysicalResourceId: physicalId,
            Status: 'FAILED',
            RequestId: event.RequestId,
            StackId: event.StackId,
            LogicalResourceId: event.LogicalResourceId,
            Reason: `failed with error ${e}`,
        };
    }
};

const onDelete = (event: CloudFormationCustomResourceDeleteEvent): CloudFormationCustomResourceResponse => {
    // If multiple resources are created this cant be hardcoded
    const physicalId = event.PhysicalResourceId;
    console.info(`delete resource ${physicalId}`);

    return {
        PhysicalResourceId: physicalId,
        Status: 'SUCCESS',
        RequestId: event.RequestId,
        StackId: event.StackId,
        LogicalResourceId: event.LogicalResourceId,
        Reason: `deleted resource successfully`,
    };
};

const onUpdate = (event: CloudFormationCustomResourceUpdateEvent): CloudFormationCustomResourceResponse => {
    // If multiple resources are created this cant be hardcoded
    const physicalId = event.PhysicalResourceId;
    console.info(`update resource ${physicalId}, no actions`);

    return {
        PhysicalResourceId: physicalId,
        Status: 'SUCCESS',
        RequestId: event.RequestId,
        StackId: event.StackId,
        LogicalResourceId: event.LogicalResourceId,
        Reason: `updated resource successfully`,
    };
};

export async function handler(event: CloudFormationCustomResourceEvent): Promise<CloudFormationCustomResourceResponse> {
    console.info({ event });

    switch (event.RequestType) {
        case 'Create':
            return await onCreate(event);
        case 'Update':
            return await onUpdate(event);
        case 'Delete':
            return onDelete(event);
    }

    throw new Error('RequestType is invalid!');
}
