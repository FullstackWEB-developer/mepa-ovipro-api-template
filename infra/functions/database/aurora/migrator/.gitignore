**/target/
**/.classpath
**/.settings/
**/.factorypath
**/.project
/bin/
