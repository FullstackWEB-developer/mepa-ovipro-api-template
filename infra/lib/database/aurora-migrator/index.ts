import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as s3 from '@aws-cdk/aws-s3';
import * as sm from '@aws-cdk/aws-secretsmanager';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as logs from '@aws-cdk/aws-logs';
import { Ec } from '@almamedia/cdk-accounts-and-environments';
import { Name } from '@almamedia/cdk-tag-and-name';
import { Duration } from '@aws-cdk/core';
import { DefaultVpc } from '../../default-resources/shared/vpc';

interface Props extends cdk.StackProps {
    migrationsBucket: s3.Bucket;
    auroraSecurityGroup: ec2.SecurityGroup;
    auroraCredentialsSecret: sm.Secret;
}

/**
 * This stack deploys Lambda DB tools for schema migration for the given Serverless cluster.
 * Required parameters: vpc, cluster and specific cluster user credentials via a Secret manager secret.
 */
export class AuroraMigratorStack extends cdk.Stack {
    public readonly handlers: lambda.IFunction[];

    constructor(scope: cdk.Construct, id: string, props: Props) {
        super(scope, id, props);

        const { auroraSecurityGroup } = props;

        const { vpc } = new DefaultVpc(this, 'DefaultVpc');
        const lambdaSecurityGroup = new ec2.SecurityGroup(this, 'MigratorLambdaSecurityGroup', {
            vpc,
        });

        auroraSecurityGroup.addIngressRule(
            lambdaSecurityGroup,
            ec2.Port.tcp(5432),
            'From Lambda migrator to Aurora cluster',
            true,
        );

        const migrator = this.createHandler(
            props,
            'Migrator',
            'fi.almamedia.ovipro.migrator.App::handleRequest',
            lambdaSecurityGroup,
            vpc,
        );

        this.handlers = [migrator];

        // Cleaner function is only included in select environments,
        // since we don't want to encourage erasing live environments like prod.
        if (Ec.isDevelopment(this)) {
            const cleaner = this.createHandler(
                props,
                'Cleaner',
                'fi.almamedia.ovipro.migrator.App::handleCleanupRequest',
                lambdaSecurityGroup,
                vpc,
            );
            this.handlers.push(cleaner);
        }
    }

    /**
     * Create Lambda handlers from the same Java package with different handler methods.
     * @param props Stack props.
     * @param id Lambda id.
     * @param handler Java Lambda handler method reference.
     * @param lambdaSecurityGroup Security group for all lambdas in this stack.
     */
    private createHandler(
        props: Props,
        id: string,
        handler: string,
        lambdaSecurityGroup: ec2.SecurityGroup,
        vpc: ec2.IVpc,
    ) {
        const { migrationsBucket, auroraCredentialsSecret } = props;

        const fn = new lambda.Function(this, id, {
            functionName: Name.withProject(this, id),
            handler,
            code: lambda.Code.fromAsset('functions/database/aurora/migrator/target/migrator-lambda-package.zip'),
            description: 'Database schema migration/cleaner utility: ' + Name.withProject(this, id),
            runtime: lambda.Runtime.JAVA_11,
            memorySize: 512,
            timeout: Duration.seconds(599),
            environment: {
                POWERTOOLS_SERVICE_NAME: Name.withProject(this, id),
                POWERTOOLS_METRICS_NAMESPACE: Name.withProject(this, handler),
                BUCKET_NAME: migrationsBucket.bucketName,
                SECRET_ARN: auroraCredentialsSecret.secretArn,
            },
            // Enable X-ray
            tracing: lambda.Tracing.ACTIVE,
            vpc,
            logRetention: logs.RetentionDays.ONE_MONTH,
            securityGroups: [lambdaSecurityGroup],
        });

        migrationsBucket.grantRead(fn);
        auroraCredentialsSecret.grantRead(fn);
        return fn;
    }
}
