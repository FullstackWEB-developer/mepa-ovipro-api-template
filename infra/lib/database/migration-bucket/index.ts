import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import { Name } from '@almamedia/cdk-tag-and-name';
import { Ac, Ec } from '@almamedia/cdk-accounts-and-environments';
import { pascalCase } from 'change-case';
import { RemovalPolicy } from '@aws-cdk/core';

export class MigrationBucketStack extends cdk.Stack {
    public readonly s3Bucket: s3.Bucket;

    constructor(scope: cdk.Construct, id: string, props: cdk.StackProps) {
        super(scope, id, props);

        const name = `ovipro${Ac.getType(this)}${pascalCase(Ec.getName(this))}migrations`;

        const s3Bucket = new s3.Bucket(this, Name.withProject(this, name), {
            bucketName: name.toLowerCase(),
            blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
            encryption: s3.BucketEncryption.S3_MANAGED,
            autoDeleteObjects: !Ec.isStable(this),
            removalPolicy: Ec.isStable(this) ? RemovalPolicy.RETAIN : RemovalPolicy.DESTROY,
        });

        this.s3Bucket = s3Bucket;
    }
}
