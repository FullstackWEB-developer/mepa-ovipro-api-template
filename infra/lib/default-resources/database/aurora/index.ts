import * as cdk from '@aws-cdk/core';
import * as rds from '@aws-cdk/aws-rds';
import { OviproSharedResource } from '../../../utils/shared-resources/OviproSharedResource';
import { SharedResourceType } from '../../../utils/shared-resources/types';

export class DefaultAuroraCluster extends cdk.Construct {
    public readonly database: rds.IServerlessCluster;

    /** Creates a Serverless Aurora database */
    constructor(scope: cdk.Construct, id: string) {
        super(scope, id);

        /**
         * Import existing RDS cluster's identifier
         */
        const sharedResource = new OviproSharedResource(this, 'SharedResource');
        const clusterIdentifier = sharedResource.import(SharedResourceType.DATABASE_CLUSTER_IDENTIFIER);

        const database = rds.ServerlessCluster.fromServerlessClusterAttributes(scope, 'DefaultServerlessCluster', {
            clusterIdentifier,
        });

        this.database = database;
    }
}
