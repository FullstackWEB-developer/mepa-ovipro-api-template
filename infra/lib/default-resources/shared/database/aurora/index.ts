import * as cdk from '@aws-cdk/core';
import * as rds from '@aws-cdk/aws-rds';
import { OviproEnvironmentSharedResource } from '../../../../utils/shared-resources/OviproEnvironmentSharedResource';
import { SharedResourceType } from '../../../../utils/shared-resources/types';

export class DefaultAuroraCluster extends cdk.Construct {
    public readonly database: rds.IServerlessCluster;

    /** Creates a Serverless Aurora database */
    constructor(scope: cdk.Construct, id: string) {
        super(scope, id);

        /**
         * Import existing RDS cluster's identifier
         */
        const sharedResource = new OviproEnvironmentSharedResource(this, 'SharedResource');
        const clusterIdentifier = sharedResource.import(SharedResourceType.DATABASE_CLUSTER_IDENTIFIER);

        const database = rds.ServerlessCluster.fromServerlessClusterAttributes(this, 'DefaultServerlessCluster', {
            clusterIdentifier,
        });

        this.database = database;
    }
}
