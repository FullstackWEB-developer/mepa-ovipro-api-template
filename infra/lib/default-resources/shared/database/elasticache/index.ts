import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import { OviproSharedResource } from '../../../../utils/shared-resources/OviproSharedResource';
import { SharedResourceType } from '../../../../utils/shared-resources/types';

interface Props extends cdk.StackProps {
    vpc: ec2.Vpc;
}

export class DefaultElasticCache extends cdk.Stack {
    public readonly endpointAddress: string;
    public readonly endpointPort: string;

    constructor(scope: cdk.Construct, id: string, props: Props) {
        super(scope, id, props);

        /**
         * Import shared cluster's address and port
         *
         * Importing cache cluster as a CDK-resource is not possible, unless either
         * we or aws creates a class where lookup-function is implemented
         */
        const sharedResource = new OviproSharedResource(this, 'SharedResource');
        const endpointAddress = sharedResource.import(SharedResourceType.CACHE_CLUSTER_ENDPOINT_ADDRESS);
        const endpointPort = sharedResource.import(SharedResourceType.CACHE_CLUSTER_ENDPOINT_PORT);

        this.endpointAddress = endpointAddress;
        this.endpointPort = endpointPort;
    }
}
