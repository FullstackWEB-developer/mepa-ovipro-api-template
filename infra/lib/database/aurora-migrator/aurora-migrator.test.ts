import { Sc } from '@almamedia/cdk-accounts-and-environments';
import { SynthUtils } from '@aws-cdk/assert';
import { Name } from '@almamedia/cdk-tag-and-name';
import { createCdkTestContext } from '../../../__test__/context';
import { AuroraClusterStack } from '../../database/aurora';
import { AuroraMigratorStack } from '../aurora-migrator';
import { MigrationBucketStack } from '../migration-bucket';

test('Aurora Migrator', () => {
    // GIVEN
    const scope = createCdkTestContext();

    const auroraStack = new AuroraClusterStack(scope, 'AuroraClusterStack', {
        stackName: Name.stack(scope, 'AuroraClusterStack'),
        ...Sc.defineProps(scope, 'Serverless Aurora cluster for OviPRO infrastructure'),
    });

    const { s3Bucket: migrationsBucket } = new MigrationBucketStack(scope, 'MigrationBucketStack', {
        stackName: Name.stack(scope, 'MigrationBucketStack'),
        ...Sc.defineProps(scope, 'Migration bucket for OviPRO infrastructure'),
    });

    const stack = new AuroraMigratorStack(scope, 'AuroraMigratorStack', {
        stackName: Name.stack(scope, 'AuroraMigratorStack'),
        ...Sc.defineProps(scope, 'Serverless Aurora cluster migrator for OviPRO infrastructure'),
        auroraCredentialsSecret: auroraStack.auroraMigratorCredentialsSecret,
        auroraSecurityGroup: auroraStack.auroraSecurityGroup,
        migrationsBucket,
    });
    // THEN

    // no need to test for assetparameter hashes
    expect.addSnapshotSerializer({
        test: (val: string) => typeof val === 'string' && !!val.match(/AssetParameters([A-Fa-f0-9]{64})(\w+)/),
        print: () => '"AssetParameters-[HASH REMOVED]"',
    });
    expect.addSnapshotSerializer({
        test: (val: string) =>
            typeof val === 'string' && !!val.match(/(\w+) for asset\s?(version)?\s?"([A-Fa-f0-9]{64})"/),
        print: (val) =>
            typeof val === 'string'
                ? '"' + val.replace(/(\w+ for asset)\s?(version)?\s?"([A-Fa-f0-9]{64})"/, '$1 [HASH REMOVED]') + '"'
                : '',
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
