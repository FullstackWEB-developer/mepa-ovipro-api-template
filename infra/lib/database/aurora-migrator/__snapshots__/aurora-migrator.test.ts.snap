// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aurora Migrator 1`] = `
Object {
  "Description": "Mock1 Environment: Serverless Aurora cluster migrator for OviPRO infrastructure",
  "Parameters": Object {
    "AssetParameters-[HASH REMOVED]": Object {
      "Description": "Artifact hash for asset [HASH REMOVED]",
      "Type": "String",
    },
    "AssetParameters-[HASH REMOVED]": Object {
      "Description": "S3 bucket for asset [HASH REMOVED]",
      "Type": "String",
    },
    "AssetParameters-[HASH REMOVED]": Object {
      "Description": "S3 key for asset [HASH REMOVED]",
      "Type": "String",
    },
    "AssetParameters-[HASH REMOVED]": Object {
      "Description": "Artifact hash for asset [HASH REMOVED]",
      "Type": "String",
    },
    "AssetParameters-[HASH REMOVED]": Object {
      "Description": "S3 bucket for asset [HASH REMOVED]",
      "Type": "String",
    },
    "AssetParameters-[HASH REMOVED]": Object {
      "Description": "S3 key for asset [HASH REMOVED]",
      "Type": "String",
    },
  },
  "Resources": Object {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters-[HASH REMOVED]",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters-[HASH REMOVED]",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters-[HASH REMOVED]",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MigratorB44932C8": Object {
      "DependsOn": Array [
        "MigratorServiceRoleDefaultPolicy8BA0D813",
        "MigratorServiceRoleE35D46BF",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters-[HASH REMOVED]",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters-[HASH REMOVED]",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters-[HASH REMOVED]",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "Database schema migration/cleaner utility: OviproMock1Migrator",
        "Environment": Object {
          "Variables": Object {
            "BUCKET_NAME": Object {
              "Fn::ImportValue": "Ovipro-Environment-Mock1-MigrationBucketStack:MigrationBucketStackExportsOutputRefOviproMock1OvipromockMock1migrationsD48EF7328136446F",
            },
            "POWERTOOLS_METRICS_NAMESPACE": "OviproMock1FiAlmamediaOviproMigratorAppHandleRequest",
            "POWERTOOLS_SERVICE_NAME": "OviproMock1Migrator",
            "SECRET_ARN": Object {
              "Fn::ImportValue": "Ovipro-Environment-Mock1-AuroraClusterStack:AuroraClusterStackExportsOutputRefMock1DbCredMigratorSecretFD4A1AB6DAA1A8F1",
            },
          },
        },
        "FunctionName": "OviproMock1Migrator",
        "Handler": "fi.almamedia.ovipro.migrator.App::handleRequest",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "MigratorServiceRoleE35D46BF",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Timeout": 599,
        "TracingConfig": Object {
          "Mode": "Active",
        },
        "VpcConfig": Object {
          "SecurityGroupIds": Array [
            Object {
              "Fn::GetAtt": Array [
                "MigratorLambdaSecurityGroup4005CC4E",
                "GroupId",
              ],
            },
          ],
          "SubnetIds": Array [
            "p-12345",
            "p-67890",
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "MigratorLambdaSecurityGroup4005CC4E": Object {
      "Properties": Object {
        "GroupDescription": "TestContextProvider/AuroraMigratorStack/MigratorLambdaSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "MigratorLambdaSecurityGroupTestContextProviderAuroraClusterStackServerlessClusterSecurityGroupB8FC79935432to21D3C046": Object {
      "Properties": Object {
        "Description": "From Lambda migrator to Aurora cluster",
        "FromPort": 5432,
        "GroupId": Object {
          "Fn::ImportValue": "Ovipro-Environment-Mock1-AuroraClusterStack:AuroraClusterStackExportsOutputFnGetAttServerlessClusterSecurityGroup5151F34FGroupId11D96200",
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "MigratorLambdaSecurityGroup4005CC4E",
            "GroupId",
          ],
        },
        "ToPort": 5432,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "MigratorLogRetentionEDE7FEB6": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "MigratorB44932C8",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "MigratorServiceRoleDefaultPolicy8BA0D813": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::ImportValue": "Ovipro-Environment-Mock1-MigrationBucketStack:MigrationBucketStackExportsOutputFnGetAttOviproMock1OvipromockMock1migrationsD48EF732Arn9D99CA52",
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::ImportValue": "Ovipro-Environment-Mock1-MigrationBucketStack:MigrationBucketStackExportsOutputFnGetAttOviproMock1OvipromockMock1migrationsD48EF732Arn9D99CA52",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::ImportValue": "Ovipro-Environment-Mock1-AuroraClusterStack:AuroraClusterStackExportsOutputRefMock1DbCredMigratorSecretFD4A1AB6DAA1A8F1",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MigratorServiceRoleDefaultPolicy8BA0D813",
        "Roles": Array [
          Object {
            "Ref": "MigratorServiceRoleE35D46BF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MigratorServiceRoleE35D46BF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
